ee/server/scim/users.go

```diff
@@ -0,0 +1,455 @@
+package scim
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/elimity-com/scim"
+	"github.com/elimity-com/scim/errors"
+	"github.com/elimity-com/scim/optional"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/scim2/filter-parser/v2"
+)
+
+const (
+	// Common attributes: https://datatracker.ietf.org/doc/html/rfc7643#section-3.1
+	externalIdAttr = "externalId"
+
+	// User attributes: https://datatracker.ietf.org/doc/html/rfc7643#section-4.1
+	userNameAttr   = "userName"
+	nameAttr       = "name"
+	givenNameAttr  = "givenName"
+	familyNameAttr = "familyName"
+	activeAttr     = "active"
+	emailsAttr     = "emails"
+)
+
+type UserHandler struct {
+	ds     fleet.Datastore
+	logger kitlog.Logger
+}
+
+// Compile-time check
+var _ scim.ResourceHandler = &UserHandler{}
+
+func NewUserHandler(ds fleet.Datastore, logger kitlog.Logger) scim.ResourceHandler {
+	return &UserHandler{ds: ds, logger: logger}
+}
+
+func (u *UserHandler) Create(r *http.Request, attributes scim.ResourceAttributes) (scim.Resource, error) {
+	// Check for userName uniqueness
+	userName, err := getRequiredResource[string](attributes, userNameAttr)
+	if err != nil {
+		level.Error(u.logger).Log("msg", "failed to get userName", "err", err)
+		return scim.Resource{}, err
+	}
+	_, err = u.ds.ScimUserByUserName(r.Context(), userName)
+	if !fleet.IsNotFound(err) {
+		level.Info(u.logger).Log("msg", "user already exists", userNameAttr, userName)
+		return scim.Resource{}, errors.ScimErrorUniqueness
+	}
```

Comment:
Double-checking, do you mean to treat `err == nil` the same as `err !=nil && !fleet.IsNotFound(err)`? In the latter case, I think it wouldn't necessarily be true that the user already exists if the error is due to some other reason and we would lose that error trace.